Adding starter example blog post to website (#39)
risc0/website: blog/2022-06-27-starter/index.md

---
slug: starter
title: Understanding the Rust starter project
authors: [elakn]
tags: [zk, zkvm, rust]
---

> In this post, we introduce a simple Rust code project written for the RISC Zero zero-knowledge virtual machine (zkVM). In this post, We won't show you how to write a full program, but reading this should help you get acquainted with writing code for the zkVM. Although code has been omitted for simplicity, we strongly recommend reading the following explanation in conjunction with the [RISC Zero Rust starter example](https://github.com/risc0/risc0-rust-starter) repository.

Reviewer: ```suggestion
In this post, we introduce a simple Rust code project written for the RISC Zero zero-knowledge virtual machine (zkVM). In this post, we won't show you how to write a full program, but reading this should help you get acquainted with writing code for the zkVM. Although code has been omitted for simplicity, we strongly recommend reading the following explanation in conjunction with the [RISC Zero Rust starter example](https://github.com/risc0/risc0-rust-starter) repository.
```

By reading this post, you should learn, at a high level:
* How RISC Zero can be used to convince someone that code has executed on the zkVM
* How we make programs and variables available to the zkVM guest
* What zkVM guest programs can do to publicly share computation results
* How this code example can be adapted to send to a real recipient

# A program to multiply two numbers in the zkVM

> In the Rust starter program, we demonstrate how to multiply two numbers and share their product with a recipient without revealing what the two factors are. The RISC Zero zkVM provides a cryptographically strong argument that we performed the computation. By sharing the computational result from the zkVM, we can tie our computational results to our argument of correct program execution.

Reviewer: ```suggestion
In the Rust starter program, we demonstrate how to multiply two numbers and share their product without revealing what the two factors are. The RISC Zero zkVM provides a cryptographically strong argument that we performed the computation, which is witnessed by a verifiable "receipt." By sharing the receipt, we can convince a skeptical third party that we ran the computation faithfully and that the output is honest.
```

# Differentiating host and guest code

> The zkVM runs as a guest on our host system and is interacted with through code that runs on the host. Most code written for the guest zkVM lives in [`methods/guest/src/bin/multiply.rs`](https://github.com/risc0/risc0-rust-starter/blob/main/methods/guest/src/bin/multiply.rs). The main function for our host program code  lives in [`starter/src/main.rs`](https://github.com/risc0/risc0-rust-starter/blob/main/starter/src/main.rs).

Reviewer: > The zkVM runs as a guest on our host system and is interacted with through code that runs on the host.
Not easy to parse

Author: Substituting "The host driver program runs the guest zkVM. "

> When the host code executes, it creates and runs a prover that is responsible for all guest zkVM interactions. The prover runs an ELF binary of the zkVM guest code. After the guest code has executed, the prover returns a [computational receipt](https://www.risczero.com/docs/explainers/proof-system/what_is_a_receipt). In our example, these are accomplished with the following line in the `starter/src/main.rs` host source code:

Reviewer: > When the host code executes, it creates and runs a prover that is responsible for all guest zkVM interactions.

The working here is imprecise; on the one hand, the host library provides a `Prover` type, and the host-side code does indeed create an instance of this type. But on the other hand, the RISC Zero Prover is a component shipped by RISC Zero, and is not created by the host.

> the prover returns a [computational receipt](https://www.risczero.com/docs/explainers/proof-system/what_is_a_receipt).

We could just say "receipt," especially since it links to the explainer

Author: The intent is "creates an instance of an object of type Prover" -- adding the relevant line for clarity. Agreed on receipt language.

```
let receipt = prover.run().unwrap();
```

# Understanding the prover

When a prover runs code on the zkVM, three important things happen:

> 1. The ELF binary execution produces an [execution trace](https://www.risczero.com/docs/explainers/proof-system/what_is_a_trace). If valid, the trace generates a `cryptographic seal` that shows the execution followed RISC-V rules.

Reviewer: ```suggestion
1. The ELF binary execution produces an [execution trace](https://www.risczero.com/docs/explainers/proof-system/what_is_a_trace). This trace is used to generate a `cryptographic seal` whose validity implies the execution followed RISC-V rules.
```

> 2. A hash of the binary called the `method ID` is appended to the receipt. Later, the recipient can check this value and know which binary was executed.

Reviewer: > Later, the recipient can check this value and know which binary was executed.

I don't think the receipt checks this value; rather, I think the verifier compares the receipt to this value as part of its verification procedure

> 3. Any values the guest shares are written to the `journal`.

Reviewer: Not true; the guest can share data with the host using `env::write()` (which does not get written to the journal)

> The method ID and a hash of the journal are included in the cryptographic seal so that the recipient will know if they have been altered.

Reviewer: In the current API, the method ID is not included anywhere in the receipt: not in the journal, not in the seal

Reviewer 2: I do believe that the MethodID is used in generating the hash that becomes the seal. But perhap we could word this a little differently to reflect that.

Reviewer 2: @intoverflow is right, it's not directly part of the seal, we only pass the MethodID when calling the prover so that we can do an internal verify. I mean we do eventually compare bits of the seal with the MethodID in verify.
